src/Server/HTTPHandler.cpp

```diff
@@ -727,6 +732,16 @@ void HTTPHandler::processQuery(
         return false;
     };
 
+    if (params.has("role"))
+    {
+        auto role_name = params.get("role");
+        auto role_id = context->getAccessControl().getID<Role>(role_name);
+        if (context->getUser()->granted_roles.isGranted(role_id))
+          context->setCurrentRoles(std::vector{role_id});
+        else
+          throw Exception(ErrorCodes::UNKNOWN_ROLE, "Role {} does not exist or not granted to the current user", role_name);
```

Comment:
If the role doesn't exist then `getID<Role>()` will throw before this line. So this error message should be probably only about grants.

Reply:
Yes, you are right; this one probably should be ACCESS_DENIED then.

A question: If we call `context->setCurrentRoles` with a non-granted role, it is no-op without a failure; the error message is then not very specific, as it does not set the role, and just skips it. Then, if we are querying a table for which you need a specific grant, the query results in an error without a hint that the role is incorrect, saying that you need a grant instead (a standard one, as if you did not provide the role at all). 

Is it intended behavior? That's why I do this `isGranted` check here, so the error message is a bit clearer.

Reply:
Thanks for pointing this out. I pushed a fix to use `SET_NON_GRANTED_ROLE` instead.